consensus/src/liveness/leader_reputation.rs

```diff
@@ -202,18 +187,186 @@ impl ReputationHeuristic for ActiveInactiveHeuristic {
+        )
+    }
+
+    pub fn count_proposals(&self, epoch: u64, history: &[NewBlockEvent]) -> HashMap<Author, i32> {
+        Self::history_iter(history, epoch, self.proposer_window_size).fold(
+            HashMap::new(),
+            |mut map, meta| {
+                let count = map.entry(meta.proposer()).or_insert(0);
+                *count += 1;
+                map
+            },
+        )
+    }
+
+    pub fn count_failed_proposals(
+        &self,
+        epoch: u64,
+        candidates: &[Author],
+        history: &[NewBlockEvent],
+    ) -> HashMap<Author, i32> {
+        Self::history_iter(history, epoch, self.proposer_window_size).fold(
+            HashMap::new(),
+            |mut map, meta| {
+                match Self::indices_to_validators(candidates, meta.failed_proposer_indices()) {
+                    Ok(failed_proposers) => {
+                        for &failed_proposer in failed_proposers {
+                            let count = map.entry(failed_proposer).or_insert(0);
+                            *count += 1;
+                        }
+                    }
+                    Err(msg) => {
+                        warn!(
+                            "Failed proposer conversion from indices failed at epoch {}, round {}: {}",
+                            meta.epoch(),
+                            meta.round(),
+                            msg
+                        )
+                    }
+                }
+                map
+            },
+        )
+    }
+}
 
-        COMMITTED_PROPOSALS_IN_WINDOW.set(committed_proposals as i64);
-        COMMITTED_VOTES_IN_WINDOW.set(committed_votes as i64);
-        LEADER_REPUTATION_WINDOW_SIZE.set(history.len() as i64);
+/// If candidate appear in the history, it's assigned active_weight otherwise inactive weight.
+pub struct ActiveInactiveHeuristic {
+    author: Author,
+    active_weight: u64,
+    inactive_weight: u64,
+    aggregation: NewBlockEventAggregation,
+}
+
+impl ActiveInactiveHeuristic {
+    pub fn new(
+        author: Author,
+        active_weight: u64,
+        inactive_weight: u64,
+        window_size: usize,
+    ) -> Self {
+        Self {
+            author,
+            active_weight,
+            inactive_weight,
+            aggregation: NewBlockEventAggregation::new(window_size, window_size),
+        }
+    }
+}
+
+impl ReputationHeuristic for ActiveInactiveHeuristic {
+    fn get_weights(
+        &self,
+        epoch: u64,
+        candidates: &[Author],
+        history: &[NewBlockEvent],
+    ) -> Vec<u64> {
+        let votes = self.aggregation.count_votes(epoch, candidates, history);
+        let proposals = self.aggregation.count_proposals(epoch, history);
+
+        COMMITTED_PROPOSALS_IN_WINDOW.set(*proposals.get(&self.author).unwrap_or(&0) as i64);
+        COMMITTED_VOTES_IN_WINDOW.set(*votes.get(&self.author).unwrap_or(&0) as i64);
+        LEADER_REPUTATION_WINDOW_SIZE.set(proposals.values().sum::<i32>() as i64);
```

Comment:
isn't it self.aggregation.proposer_window_size?

Reply:
Maybe we should rename the counter to be clearer what it measures?

Reply:
changing to LEADER_REPUTATION_HISTORY_SIZE

it keeps the same meaning as before : Total number of new block events in the current reputation window

we wanted the effective number - to make sure we are fetching enough of the history. 
number that is in configuration is not really useful, especially after we move this on-chain.